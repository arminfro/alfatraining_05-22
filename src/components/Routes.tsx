import { ReactElement } from "react";
import { Navigate, Route, Routes as RRDRoutes } from "react-router-dom";

import ClassCounter from "./ClassCounter";
import FunctionalCounter from "./FunctionalCounter";
import ProjectCreate from "./ProjectCreate";
import ProjectDetails from "./ProjectDetails";
import ProjectEdit from "./ProjectEdit";
import ProjectList from "./ProjectList";

function Routes(): ReactElement {
  return (
    <RRDRoutes>
      <Route path="/functional-counter" element={<FunctionalCounter />} />
      <Route path="/class-counter" element={<ClassCounter />} />
      <Route path="/projects/new" element={<ProjectCreate />} />
      <Route path="/projects/:projectId/edit" element={<ProjectEdit />} />
      <Route path="/projects/:projectId" element={<ProjectDetails />} />
      <Route path="/projects" element={<ProjectList />} />
      <Route path="/home" element={<p>Home</p>} />
      <Route path="/" element={<Navigate to="/home" />} />
    </RRDRoutes>
  );
}

export default Routes;
