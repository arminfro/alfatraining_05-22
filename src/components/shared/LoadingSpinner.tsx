import { ReactElement } from "react";

interface Props {
  name?: string;
}

export default function LoadingSpinner(props: Props): ReactElement {
  return (
    <div className="section">
      <div className="container">
        <div className="columns">
          <div className="column is-10 is-offset-1">
            <div className="loader-wrapper">
              <div className="loader is-loading" />
            </div>
            <p className="pt-6 has-text-centered">
              Lade {props.name || "Daten"}
            </p>
          </div>
        </div>
      </div>
    </div>
  );
}
