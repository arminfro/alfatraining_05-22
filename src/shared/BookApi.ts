import { useState, useEffect, SetStateAction, Dispatch } from "react";
import axios, { AxiosResponse, Method } from "axios";

// local utility type
type SetState<T> = Dispatch<SetStateAction<T | undefined>>;

/*
 * Abstracts away both needs for api calls,
 * on rendering and on events / conditions
 *
 * useBookApi, hook
 * bookApi, normal function
 *
 */

/*
 * Useful for http data as a dependency in rendering
 *
 * @param method [Method], http method
 * @param path [string], relative path to baseUrl
 * @return, Response Data
 */
export function useBookApi<T>(path: string): [T | undefined, SetState<T>] {
  const [data, setData] = useState<T>();

  useEffect(() => {
    bookApi("GET", path, setData);
  }, [path]);

  return [data, setData];
}

/*
 * Useful for calls on events or in conditions
 *
 * @param method [Method], http method
 * @param path [string], relative path to baseUrl
 * @param data [function], callback, gets `response.data` as an argument
 * @param data [object], body data
 */
export function bookApi<T>(
  method: Method,
  path: string,
  callback: SetState<T>,
  data = {}
): void {
  const baseUrl = "https://api3.angular-buch.com/secure";

  axios({
    method: method,
    url: `${baseUrl}/${path}`,
    headers: { Authorization: "Bearer 1234567890" },
    data,
  }).then((response: AxiosResponse<T>) => callback(response.data));
}
