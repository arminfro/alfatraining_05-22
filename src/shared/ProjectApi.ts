import { useState, useEffect, SetStateAction, Dispatch } from "react";
import axios, { AxiosResponse, Method } from "axios";

import {
  isProject,
  isProjectArray,
  factoryRawToProject,
} from "../types/Project";

type Setter<T> = Dispatch<SetStateAction<T | undefined>>;

/*
 * Abstracts away both needs for api calls,
 * on rendering and on events / conditions
 *
 * useProjectApi, hook
 * projectApi, normal function
 *
 */

/*
 * Useful for http data as a dependency in rendering
 *
 * @param method [Method], http method
 * @param path [string], relative path to baseUrl
 * @return, Response Data
 */
export function useProjectApi<T>(path: string): [T | undefined, Setter<T>] {
  const [data, setData] = useState<T>();

  useEffect(() => {
    projectApi("GET", path, setData);
  }, [path]);

  return [data, setData];
}

/*
 * Useful for calls on events or in conditions
 *
 * @param method [Method], http method
 * @param path [string], relative path to baseUrl
 * @param data [function], callback, gets `response.data` as an argument
 * @param data [object], body data
 */
export function projectApi<T>(
  method: Method,
  path: string,
  callback: Setter<T>,
  data = {}
): void {
  const baseUrl = "/api";

  axios({
    method: method,
    url: `${baseUrl}/${path}`,
    data,
  }).then((response: AxiosResponse<T>) => callback(response.data));
}

/*
 * Axios Interceptor
 * Factory ProjectRaw to Project
 */
axios.interceptors.response.use(
  (response: AxiosResponse) => {
    if (isProject(response.data)) {
      response.data = factoryRawToProject(response.data);
    } else if (isProjectArray(response.data)) {
      response.data = response.data.map((project) =>
        factoryRawToProject(project)
      );
    }
    return response;
  },
  (error) => Promise.reject(error)
);
