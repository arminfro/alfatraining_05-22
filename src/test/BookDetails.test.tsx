import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import userEvent from "@testing-library/user-event";
import { books, matchSnapshot, render, screen } from "./utils";
import { baseUrl } from "../shared/BookApi";
import Routes from "../components/Routes";

const renderBookDetails = () =>
  render(<Routes />, { route: "/books/9783864906466" });

describe("BookDetails rendering and button clicks", () => {
  let mock: MockAdapter;

  beforeEach(() => {
    mock = new MockAdapter(axios);
    mock.onGet(`${baseUrl}/books/9783864906466`).reply(200, books[0]);
  });

  afterEach(() => mock.reset());

  test("matches snapshot", async () => {
    const view = renderBookDetails();
    await screen.findByTestId(/book-details/i);
    matchSnapshot(view);
  });

  test("renders", async () => {
    renderBookDetails();
    expect(await screen.findByTestId(/book-details/i)).toBeInTheDocument();
  });

  test('click on "Add To Cart" and stay on Book Details', async () => {
    renderBookDetails();
    userEvent.click(await screen.findByText(/Add To Cart/i));
    expect(await screen.findByTestId(/book-details/i)).toBeInTheDocument();
  });
});
