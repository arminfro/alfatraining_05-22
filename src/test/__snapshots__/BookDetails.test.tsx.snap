// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails rendering and button clicks matches snapshot 1`] = `
<body>
  <div>
    <h1
      class="title"
      data-testid="book-details"
    >
      Angular
    </h1>
    <div
      class="level"
    >
      <div
        class="level-item has-text-centered"
      >
        <div>
          <p
            class="heading"
          >
            ISBN
          </p>
          <p
            class="title"
          >
            9783864906466
          </p>
        </div>
      </div>
      <div
        class="level-item has-text-centered"
      >
        <div>
          <p
            class="heading"
          >
            published
          </p>
          <p
            class="title"
          >
            4/30/2019
          </p>
        </div>
      </div>
      <div
        class="level-item has-text-centered"
      >
        <div>
          <p
            class="heading"
          >
            Rating
          </p>
          <p
            class="title"
          >
            <i
              class="ion-star has-text-warning"
            />
            <i
              class="ion-star has-text-warning"
            />
            <i
              class="ion-star has-text-warning"
            />
            <i
              class="ion-star has-text-warning"
            />
            <i
              class="ion-star has-text-warning"
            />
          </p>
        </div>
      </div>
    </div>
    <div
      class="columns is-vcentered"
    >
      <div
        class="column is-3"
      >
        <h4
          class="subtitle"
        >
          Authoren
        </h4>
        Ferdinand Malcher, Johannes Hoppe, Danny Koppenhagen
      </div>
      <div
        class="column is-9"
      >
        <h4
          class="subtitle"
        >
          Beschreibung
        </h4>
        Die Autoren führen Sie mit einem anspruchsvollen Beispielprojekt durch die Welt von Angular. Lernen Sie Schritt für Schritt, wie Sie strukturierte und modulare Single-Page-Anwendungen entwickeln. Nach der erfolgreichen ersten Auflage wurde dieses Buch grundlegend aktualisiert und erweitert:
      Durchgängig aktualisiert auf Angular 7 und neuere Versionen, Redux mit Reactive Extensions for Angular (NgRx), Ausführliches Kapitel zu RxJS und Observables, Server-Side Rendering mit Angular Universal, HTTP-Interceptoren, Kompakter Schnelleinstieg in Angular mit Stackblitz.
      Praktisch: Der Programmcode zu jeder einzelnen Entwicklungsphase ist auf GitHub verfügbar. So können Sie alle Schritte gut nachvollziehen und auch Teile überspringen.
      Die Autoren Ferdinand Malcher, Johannes Hoppe und Danny Koppenhagen sind erfahrene Workshopleiter und internationale Konferenzsprecher. In diesem praxisorientierten Buch vermitteln sie die Erkenntnisse und Best Practices aus über 3 Jahren täglicher Arbeit mit Angular. Neben den Grundlagen werden auch fortgeschrittene Themen abgedeckt. Weitere behandelte Themen sind unter anderem:
      Testing mit Jasmine, Karma und Protractor, Angular CLI, Komponenten, Pipes und Direktiven, Modulsystem, Routing, Formularverarbeitung, Dependency Injection und Services, Ahead-of-Time-Kompilierung (AOT), Deployment, Internationalisierung (i18n), Mobile Anwendungen mit NativeScript.
      Das Buch setzt lediglich Vorkenntnisse in JavaScript, HTML und CSS voraus. Wer noch nicht mit TypeScript vertraut ist, findet in diesem Buch eine kompakte Einführung.
      Nach der Lektüre sind Sie für den Projektalltag mit Angular gewappnet und können robuste Webanwendungen mit dem Framework entwickeln. Auf der Website zum Buch werden außerdem regelmäßig Aktualisierungen und Neuigkeiten rund um Angular veröffentlicht.
      </div>
    </div>
    <div
      class="columns is-vcentered"
    >
      <img
        alt="Front Cover"
        class="image"
        src="https://api3.angular-buch.com/images/angular_auflage2.jpg"
        style="width: 140px;"
      />
      <img
        alt="Rückseite"
        class="image"
        src="https://api3.angular-buch.com/images/angular_auflage2_back.jpg"
        style="width: 140px;"
      />
      <img
        alt="Update Hinweis"
        class="image"
        src="https://api3.angular-buch.com/images/angular_auflage2_update-note.jpg"
        style="width: 140px;"
      />
      <img
        alt="Als gebundenes Buch und E-Book verfügbar"
        class="image"
        src="https://api3.angular-buch.com/images/angular_auflage2_book+ebook.jpg"
        style="width: 140px;"
      />
    </div>
    <button
      class="button m-1 is-link"
    >
      Back
    </button>
    <button
      class="button m-1 is-danger"
    >
      Delete
    </button>
    <button
      class="button m-1 is-warning"
    >
      <a
        href="/books/9783864906466/edit"
      >
        Edit
      </a>
    </button>
    <button
      class="button m-1 is-primary"
    >
      Add To Cart
    </button>
  </div>
</body>
`;
